src/DynamicRupture/Output/Base.cpp

```diff
@@ -290,6 +246,170 @@ void Base::calcFaultOutput(const OutputType type, OutputData& outputData, double
+}
+
+void Base::computeSlipAndRate(const double* tangent1,
+                              const double* tangent2,
+                              const double* strike,
+                              const double* dip) {
+  local.srS = static_cast<real>(0.0);
+  local.srD = static_cast<real>(0.0);
+
+  for (size_t i = 0; i < 3; ++i) {
+    real factorMinus = (local.faceAlignedValuesMinus[7] * tangent1[i] +
+                        local.faceAlignedValuesMinus[8] * tangent2[i]);
+
+    real factorPlus = (local.faceAlignedValuesPlus[7] * tangent1[i] +
+                       local.faceAlignedValuesPlus[8] * tangent2[i]);
+
+    local.srS += factorMinus * strike[i] - factorPlus * strike[i];
+    local.srD += factorMinus * dip[i] - factorPlus * dip[i];
+  }
+}
+
+int Base::getClosestInternalGp(int nearestGpIndex, int nPoly) {
+  int i1 = int((nearestGpIndex - 1) / (nPoly + 2)) + 1;
+  int j1 = (nearestGpIndex - 1) % (nPoly + 2) + 1;
+  if (i1 == 1) {
+    i1 = i1 + 1;
+  } else if (i1 == (nPoly + 2)) {
+    i1 = i1 - 1;
+  }
+
+  if (j1 == 1) {
+    j1 = j1 + 1;
+  } else if (j1 == (nPoly + 2)) {
+    j1 = j1 - 1;
+  }
+  return (i1 - 1) * (nPoly + 2) + j1;
+}
+
+real Base::computeRuptureVelocity(Eigen::Matrix<real, 2, 2>& jacobiT2d) {
+  auto* ruptureTime = (local.layer->var(drDescr->ruptureTime))[local.ltsId];
+  real ruptureVelocity = 0.0;
+
+  bool needsUpdate{true};
+  for (size_t point = 0; point < tensor::QInterpolated::Shape[0]; ++point) {
+    if (ruptureTime[point] == 0.0) {
+      needsUpdate = false;
+    }
+  }
+
+  if (needsUpdate) {
+    constexpr int numPoly = CONVERGENCE_ORDER - 1;
+    constexpr int numDegFr2d = (numPoly + 1) * (numPoly + 2) / 2;
+    std::array<real, numDegFr2d> projectedRT{};
+    projectedRT.fill(0.0);
+
+    std::array<real, 2 * numDegFr2d> phiAtPoint{};
+    phiAtPoint.fill(0.0);
+
+    auto chiTau2dPoints =
+        init::quadpoints::view::create(const_cast<real*>(init::quadpoints::Values));
+    auto weights = init::quadweights::view::create(const_cast<real*>(init::quadweights::Values));
+
+    auto* rt = local.layer->var(drDescr->ruptureTime);
+    for (size_t jBndGP = 0; jBndGP < misc::numberOfBoundaryGaussPoints; ++jBndGP) {
+      real chi = chiTau2dPoints(jBndGP, 0);
+      real tau = chiTau2dPoints(jBndGP, 1);
+      computeTriDubinerPolynomials(phiAtPoint.data(), chi, tau, numPoly);
+
+      for (size_t d = 0; d < numDegFr2d; ++d) {
+        projectedRT[d] += weights(jBndGP) * rt[local.ltsId][jBndGP] * phiAtPoint[d];
+      }
+    }
+
+    auto m2inv =
+        seissol::init::M2inv::view::create(const_cast<real*>(seissol::init::M2inv::Values));
+    for (size_t d = 0; d < numDegFr2d; ++d) {
+      projectedRT[d] *= m2inv(d, d);
+    }
+
+    auto nearestGpIndex = getClosestInternalGp(local.nearestGpIndex, numPoly);
+    real chi = chiTau2dPoints(nearestGpIndex, 0);
+    real tau = chiTau2dPoints(nearestGpIndex, 1);
```

Comment:
Building on Thomas' idea, we would replace this with something like `chi, tau = movePointCloserToCenter(chiTau2dPoints(local.nearestGpIndex, 0), chiTau2dPoints(local.nearestGpIndex, 1)`.

Reply:
I was looking into how one could easily do that. A modified warp and blend strategy as e.g. in Hesthaven & Warburton (Fig 6.5) might be a good idea.

Reply:
```
import numpy as np
import matplotlib.pyplot as plt


def equidistant_points(N, corners):
    M = (N+2) * (N+1) // 2
    cartesian_coords = np.zeros((2, M))
    barycentric_coords = np.zeros((3, M))
    k = 0
    for i in range(N+1):
        for j in range(N-i+1):
            barycentric_coords[0, k] = i / N
            barycentric_coords[1, k] = j / N
            barycentric_coords[2, k] = 1 - barycentric_coords[0, k] - barycentric_coords[1, k]
            cartesian_coords[:,k] = np.dot(corners, barycentric_coords[:,k])
            k += 1
    return cartesian_coords, barycentric_coords

def warp_and_blend(barycentric_coords, corners):
    w = lambda x: -np.sin(0.5*x*np.pi) / (1 - x**2)
    result = np.zeros((3, 2, barycentric_coords.shape[1]))

    def w_a_b(barycentric_coords, corners, i, j):
        warp = w(barycentric_coords[i,:] - barycentric_coords[j,:])
        blend = 4 * barycentric_coords[i,:] * barycentric_coords[j,:]
        vector = corners[:,i] - corners[:,j]
        vector = vector / np.linalg.norm(vector)
        return np.multiply(blend, np.outer(warp, vector).T)

    result[0,:,:] = w_a_b(barycentric_coords, corners, 2, 1)
    result[1,:,:] = w_a_b(barycentric_coords, corners, 0, 2)
    result[2,:,:] = w_a_b(barycentric_coords, corners, 1, 0)

    return result

corners = np.array([[0, 1, 0.1], [0.1, 0.2, 1]])
points, barycentric_coords = equidistant_points(10, corners)
warp = warp_and_blend(barycentric_coords, corners)
fig, axes = plt.subplots(2, 4)
axes[0, 0].scatter(points[0,:], points[1,:])
axes[0, 0].plot([corners[0,0], corners[0,1], corners[0,2], corners[0,0]], [corners[1,0], corners[1,1], corners[1,2], corners[1,0]])
axes[0, 1].scatter(points[0,:], points[1,:], c=barycentric_coords[0,:])
axes[0, 1].set_title("l_0")
axes[0, 2].scatter(points[0,:], points[1,:], c=barycentric_coords[1,:])
axes[0, 2].set_title("l_1")
axes[0, 3].scatter(points[0,:], points[1,:], c=barycentric_coords[2,:])
axes[0, 3].set_title("l_2")

for k in range(3):
    axes[1, k+1].quiver(points[0,:], points[1,:], warp[k, 0,:], warp[k, 1,:])
axes[1, 0].quiver(points[0,:], points[1,:], np.sum(warp[:, 0,:], axis=0), np.sum(warp[:,1,:], axis=0))

for ax in axes.flatten():
    ax.set_aspect(1)
plt.show()
```
![figure](https://user-images.githubusercontent.com/24592027/159662028-04ab880d-81d6-4eca-9db3-3dffd79e4f91.png)

Reply:
at the end, I m wondering if it is not overkilling. We could just try to implement it in the most simple way without any correction, and see if with Dunavant quadrature rule we also have artifacts in the rupture speed pattern. Maybe most of the artifacts are due to the uneven distribution of Strout quadrature points?